src/bottom-bar/use-on-complete-callback.js

```diff
@@ -0,0 +1,166 @@
+import { useAlert } from '@dhis2/app-runtime'
+import i18n from '@dhis2/d2-i18n'
+import { useIsMutating } from '@tanstack/react-query'
+import { useSetRightHandPanel } from '../right-hand-panel/index.js'
+import {
+    selectors,
+    useConnectionStatus,
+    useDataSetId,
+    useImperativeCancelCompletionMutation,
+    useImperativeValidate,
+    useMetadata,
+    useSetFormCompletionMutation,
+    useSetFormCompletionMutationKey,
+    validationResultsSidebarId,
+} from '../shared/index.js'
+
+const validationFailedMessage = i18n.t(
+    "Couldn't validate the form, please try again later"
+)
+const validationFailedWarningMessage = i18n.t(
+    "Couldn't validate the form. This does not effect form completion!"
+)
+const completingFormFailedMessage = i18n.t(
+    "Couldn't validate the form, please try again later"
+)
+
+function hasViolations(commentRequiredViolations, validationRuleViolations) {
+    return (
+        commentRequiredViolations.length ||
+        Object.keys(validationRuleViolations).length
+    )
+}
+
+function useOnCompleteWhenValidRequiredClick() {
+    const { mutateAsync: setFormCompletion } = useSetFormCompletionMutation()
+    const setRightHandPanel = useSetRightHandPanel()
+    const validate = useImperativeValidate()
+
+    return () =>
+        validate()
+            // Show this alert message and skip the next "then" block
+            .catch((e) => {
+                console.error(e)
+                throw new Error(validationFailedMessage)
+            })
+            // the validation request succeeded
+            .then(({ commentRequiredViolations, validationRuleViolations }) => {
+                // if the form is invalid, show the sidebar and show an
+                // alert to the user
+                if (
+                    hasViolations(
+                        commentRequiredViolations,
+                        validationRuleViolations
+                    )
+                ) {
+                    setRightHandPanel(validationResultsSidebarId)
+                    return Promise.reject(
+                        new Error(
+                            i18n.t("The form can't be completed while invalid")
+                        )
+                    )
+                }
+                // otherwise complete the form
+                else {
+                    return setFormCompletion({ completed: true }).catch((e) => {
+                        console.error(e)
+                        throw new Error(completingFormFailedMessage)
+                    })
+                }
+            })
+}
+
+function useOnCompleteWhenValidNotRequiredClick() {
+    const { mutateAsync: setFormCompletion } = useSetFormCompletionMutation()
+    const setRightHandPanel = useSetRightHandPanel()
+    const validate = useImperativeValidate()
```

Comment:
Is there a reason this needs to be imperative?

Why can't we do a regular `useQuery` here, with a `onError` ? The error is ignored below anyway, and the `setFormCompletion` does not seem to be dependant on the return-value of this?

There's some benefits to it I think, like react-query handling when the component is unmounted, so it doesn't call `onSuccess` etc.

Reply:
We might want to just keep it like this for now though...